src/base/bittorrent/customstorage.h

```diff
@@ -30,10 +30,76 @@
+                    , lt::disk_job_flags_t flags) override;
+    bool async_write(lt::storage_index_t storageIndex, const lt::peer_request &peerRequest
+                     , const char *buf, std::shared_ptr<lt::disk_observer> diskObserver
+                     , std::function<void (const lt::storage_error &)> handler, lt::disk_job_flags_t flags) override;
+    void async_hash(lt::storage_index_t storageIndex, lt::piece_index_t piece, lt::span<lt::sha256_hash> hash, lt::disk_job_flags_t flags
+                    , std::function<void (lt::piece_index_t, const lt::sha1_hash &, const lt::storage_error &)> handler) override;
+    void async_hash2(lt::storage_index_t storage, lt::piece_index_t piece, int offset, lt::disk_job_flags_t flags
+                     , std::function<void (lt::piece_index_t, const lt::sha256_hash &, const lt::storage_error &)> handler) override;
+    void async_move_storage(lt::storage_index_t storage, std::string path, lt::move_flags_t flags
+                            , std::function<void (lt::status_t, const std::string &, const lt::storage_error &)> handler) override;
+    void async_release_files(lt::storage_index_t storage, std::function<void ()> handler) override;
+    void async_check_files(lt::storage_index_t storage, const lt::add_torrent_params *resume_data
+                           , lt::aux::vector<std::string, lt::file_index_t> links
+                           , std::function<void (lt::status_t, const lt::storage_error &)> handler) override;
+    void async_stop_torrent(lt::storage_index_t storage, std::function<void ()> handler) override;
+    void async_rename_file(lt::storage_index_t storage, lt::file_index_t index, std::string name
+                           , std::function<void (const std::string &, lt::file_index_t, const lt::storage_error &)> handler) override;
+    void async_delete_files(lt::storage_index_t storage, lt::remove_flags_t options, std::function<void (const lt::storage_error &)> handler) override;
+    void async_set_file_priority(lt::storage_index_t storage, lt::aux::vector<lt::download_priority_t, lt::file_index_t> priorities
+                                 , std::function<void (const lt::storage_error &, lt::aux::vector<lt::download_priority_t, lt::file_index_t>)> handler) override;
+    void async_clear_piece(lt::storage_index_t storage, lt::piece_index_t index, std::function<void (lt::piece_index_t)> handler) override;
+    void update_stats_counters(lt::counters &counters) const override;
+    std::vector<lt::open_file_state> get_status(lt::storage_index_t index) const override;
+    void abort(bool wait) override;
+    void submit_jobs() override;
+    void settings_updated() override;
+
+private:
+    void handleCompleteFiles(libtorrent::storage_index_t storage, const QString &savePath);
+
+    std::unique_ptr<lt::disk_interface> m_nativeDiskIO;
+
+    struct StorageData
+    {
+        QString savePath;
+        lt::file_storage files;
```

Comment:
I think the link error of `file_entry::~file_entry` is fundamentally caused by `file_storage::operator=(file_storage&&)` being inlined in libtorrent. It shouldn't be. I will fix that. One reason this hasn't come up before is that the `file_storage` object isn't really meant to be copied like this. It would be safe to keep a pointer to it here instead of a copy. The `file_storage` is immutable and will stay alive as long as the torrent is alive.

I would recommend this member be changed to:

```
lt::file_storage const* files;
```

Reply:
https://github.com/arvidn/libtorrent/pull/5225

Reply:
>The `file_storage` is immutable and will stay alive as long as the torrent is alive.

Either I don't understand something, or even you have a false idea about the convenience of the libtorrent interface... 
How can you get a pointer to the internal file_storage object here?

Reply:
right now you copy `lt::storage_params::files` (or `lt::storage_params::mapped_files`). I'm suggesting you take a pointer instead.

I realize that there's also an issue in my documentation generation script that fails to include `storage_params` in the reference documentation. I'll have to address that too.

Reply:
here: https://github.com/qbittorrent/qBittorrent/pull/13395/files#diff-7fe7f0ec96fa95f6089367abeaf9e325356951b2bdab4d7469f11b468baf1d55R59

Reply:
I applied the patch against libtorrent RC_2_0, and qBittorrent compiled successfully, but crashes at launch:

```
ASSERT: "m_metricIndices.disk.numBlocksCacheHits >= 0" in file /home/zywo/qbittorrent/src/base/bittorrent/session.cpp, line 1192

Thread 1 "qbittorrent" received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff5788537 in __GI_abort () at abort.c:79
#2  0x00007ffff5ce3b51 in  () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#3  0x00007ffff5ce2fa0 in qt_assert_x(char const*, char const*, char const*, int) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#4  0x000055555560ffcb in BitTorrent::Session::initMetrics() (this=0x55555624a7a0)
    at /home/zywo/qbittorrent/src/base/bittorrent/session.cpp:1192
#5  0x000055555560ad28 in BitTorrent::Session::Session(QObject*) (this=0x55555624a7a0, parent=0x0)
    at /home/zywo/qbittorrent/src/base/bittorrent/session.cpp:511
#6  0x000055555560e374 in BitTorrent::Session::initInstance() () at /home/zywo/qbittorrent/src/base/bittorrent/session.cpp:967
#7  0x00005555555d4b49 in Application::exec(QStringList const&) (this=0x555555e956b0, params=...)
    at /home/zywo/qbittorrent/src/app/application.cpp:572
#8  0x00005555555e8b70 in main(int, char**) (argc=1, argv=0x7fffffffe1e8) at /home/zywo/qbittorrent/src/app/main.cpp:290

Reply:
>qBittorrent compiled successfully, but crashes at launch:

I encountered this issue when tested this PR. But it is unrelated.

Reply:
>right now you copy `lt::storage_params::files` (or `lt::storage_params::mapped_files`). I'm suggesting you take a pointer instead.

Yes.
But if the files have original names in this place, then I will get a pointer to `lt::storage_params::files` that will become useless after the first renaming of the file. So I have to have a copy of actual file names and duplicate every rename there.

Reply:
ah, I see

Reply:
If we had access to the corresponding torrent `(torrent_handle)` in `new_torrent()`, we could save it and access internal `file_storage` (etc.) using its interface.